New package: QGIS

ar-jan opened the pull request:
<!-- Uncomment relevant sections and delete options which are not applicable -->

#### Testing the changes
- I tested the changes in this PR: **yes**

#### New package
- This new package conforms to the [package requirements](https://github.com/void-linux/void-packages/blob/master/CONTRIBUTING.md#package-requirements): **YES**

New packages:

- QGIS
- libkml

<!-- Note: If the build is likely to take more than 2 hours, please add ci skip tag as described in
https://github.com/void-linux/void-packages/blob/master/CONTRIBUTING.md#continuous-integration
and test at least one native build and, if supported, at least one cross build.
Ignore this section if this PR is not skipping CI.
-->
<!--
#### Local build testing
- I built this PR locally for my native architecture, (ARCH-LIBC)
- I built this PR locally for these architectures (if supported. mark crossbuilds):
  - aarch64-musl
  - armv7l
  - armv6l-musl
-->


Completed: ~Depends on #31620 being updated/merged.~

ar-jan commented:
Keeping this on the [LTS branch](https://qgis.org/en/site/getinvolved/development/roadmap.html#release-schedule) for now. If/when the qscintilla PR is merged and this could get in, ideally I'd like to have both `qgis` and `qgis-lts` packages.

cinerion commented:
Tested the package in my PC, it works properly, except for one missing runtime dependency, `python3-PyQt5-location`. Once i installed it i had no issues.

ar-jan commented:
Good to know, I'll add it with a next version bump.

ar-jan commented:
Updated to 3.28.8.

ar-jan commented:
Rebased and updated to 3.28.9.

classabbyamp commented:
tested this out on x86_64. is it ready?

ar-jan commented:
Can we include both the latest stable and LTR release? Currently that would be `qgis-3.32.1` and `qgis-ltr-3.28.9`.

If yes, is there some xbps-src option to specify `qgis-ltr` as the name of the binary/package dir?

And if we go with a single version, shall we go with the latest stable? I've now tried 3.32.1 as well. Also made a few small changes to the template (compilation flags to disable pdal and grass, and removed a few makedepends which I think are not needed). Otherwise I think it's ready.

classabbyamp commented:
i think it's fine to do latest stable only

ar-jan commented:
I want to get the GRASS support functioning, but get `"Could not find GRASS 8"` in the config step. It looks like there's something wrong with how grass is currently packaged. I added a missing `vinstall grass.pc 644 usr/lib/pkgconfig` to the grass template, but notice it contains the wrong location:

``` sh
# Package Information for pkg-config
#
# See also: grass --config

prefix=/opt/grass/grass-8.2.1
exec_prefix=/opt/grass/grass-8.2.1
libdir=/opt/grass/grass-8.2.1/lib
includedir=/opt/grass/grass-8.2.1/include

Name: GRASS
Description: GRASS GIS
Version: 8.2.1
Libs: -L/opt/grass/grass-8.2.1/lib -lgrass_gmath -lgrass_gis -lgrass_datetime -lgrass_gproj -lgrass_raster -lgrass_vector -lgrass_dbmibase -lgrass_dbmiclient -lm -lz
Cflags: -I/opt/grass/grass-8.2.1/include
```

The actual path is `/opt/grass/grass-8.2`. Any ideas? Elsewhere the correct path is used, e.g. for `GISBASE` in /usr/bin/grass.

Additionally, I suppose the libraries listed in grass.pc should also be added to `shlib_provides`?

classabbyamp commented:
yeah I think we discovered grass had issues in #31687

ar-jan commented:
Yes, but it was updated to python3 and grass 8 in #40347, and the cli does seem to work. Actually QGIS does show GRASS operations in the processing toolbox and a few I tried seem to work, but I'm not sure if they're complete. Still, something is wrong.

nezos commented:
I was not monitoring the thread, so I moved independently. In case they help I attach my templates and a patch I did for resolving grass not found (it's hardlinked!):

[libpdal.txt](https://github.com/void-linux/void-packages/files/12209339/libpdal.txt)
[libgeotiff.txt](https://github.com/void-linux/void-packages/files/12209348/libgeotiff.txt)
[libgdal.txt](https://github.com/void-linux/void-packages/files/12209342/libgdal.txt)
[grass.txt](https://github.com/void-linux/void-packages/files/12209345/grass.txt)
[qgis.txt](https://github.com/void-linux/void-packages/files/12209346/qgis.txt)
[FindGRASS.patch.txt](https://github.com/void-linux/void-packages/files/12209354/FindGRASS.patch.txt)

The problem with grass that it compiles well with qgis but when I try to install qgis it cannot find the grass shared libraries. The reason is that in the grass project, they name libraries as VERSION.so and in voidlinux they need to be .so.VERSION . I couldn't find a solution as many makefiles need to be patched and this would complicate things alot. If you have a grass package that provides shlibs then you have the solution!

For the LTR I don't know how useful it is, especially considering that each new version usually changes something at the project files created by the user...

ar-jan commented:
Thanks for chiming in, I'll see if your examples help out here.

> For the LTR I don't know how useful it is, especially considering that each new version usually changes something at the project files created by the user...

Yeah, that's exactly why I think it's useful to have. If you collaborate with someone they may not have the the latest version of QGIS available, but everyone can generally use the current LTR. But personally I'm also fine with the latest version so let's start with that.

> updating wxPython4 to the latest version would make the GUI of grass work
Yeah, but then someone needs to update wxPython for all of Void first which is not trivial. IIRC it requires upgrading wxWidgets, which is/was incompatible with other packages.

ar-jan commented:
I thought [this branch](https://github.com/ar-jan/void-packages/commits/qgis-proj) _should_ solve the grass issues, I fixed the install location for pkg-config and added a -devel subpackage (I pushed to another branch to avoid needless CI build here). The QGIS build configuration step can now find grass as well. But when I install qgis I still get this:

```
qgis-3.32.3_1: broken, unresolvable shlib `libgrass_dbmibase.8.3.so'
qgis-3.32.3_1: broken, unresolvable shlib `libgrass_dbmiclient.8.3.so'
qgis-3.32.3_1: broken, unresolvable shlib `libgrass_gis.8.3.so'
qgis-3.32.3_1: broken, unresolvable shlib `libgrass_gproj.8.3.so'
qgis-3.32.3_1: broken, unresolvable shlib `libgrass_imagery.8.3.so'
qgis-3.32.3_1: broken, unresolvable shlib `libgrass_raster.8.3.so'
qgis-3.32.3_1: broken, unresolvable shlib `libgrass_vector.8.3.so'
Transaction aborted due to unresolved shlibs.
```

Do they have to be located in `/usr/lib` to be found? Any clue, @classabbyamp?

0x5c commented:
> Do they have to be located in `/usr/lib` to be found?

In what dir are they located?

ar-jan commented:
`/opt/grass83/lib`.

0x5c commented:
That's likely to be it, shlibs should be in the usual `/usr/lib`
Do we know why the prefix got changed to `/opt/grass` in e8abe0a01b81041e245e399730ab9b61141ceb89?

0x5c commented:
Debian seems to use the normal prefix https://packages.debian.org/sid/amd64/grass-core/filelist
https://salsa.debian.org/debian-gis-team/grass/-/tree/master/debian

ar-jan commented:
I'm stuck. I've changed it to install to `/usr/lib/grass83`. I'm not quite sure why `/opt` was picked, but before that it was already strange to first install to `/usr/share` and then move the whole install dir to `/usr/lib` anyway. Grass has `--includedir=DIR` and `--libdir=DIR` options but I couldn't get those options to work, setting it to e.g. `--includedir=/usr/include` appears to do nothing.

The library paths for `grass.pc` had to be fixed, but I'm still getting the `unresolvable shlib` problem. Grass compile and install appears to work fine. QGIS can be compiled with grass and the configure phase reports finding grass, but then when actually installing qgis the grass libraries can't be found (if I remove grass-devel from qgis makedepends it can be installed).

If I compile QGIS with grass-devel in makedepends, it starts like this:
> ./xbps-src pkg -C qgis
> Package grass was not found in the pkg-config search path.
> Perhaps you should add the directory containing `grass.pc'
> to the PKG_CONFIG_PATH environment variable
> No package 'grass' found

But despite that, qgis__do_configure.log then says:
> -- Found GRASS 8: /usr/lib/grass83 (8.3.0, off_t size = )

(Maybe because of passing setting the location with configure arg `GRASS_PREFIX8`?)

Without grass-devel, I get the `PKG_CONFIG_PATH` warning; when I install the grass-devel built in that same branch, I get this:

> pkg-config --variable=prefix grass
> /usr/lib/grass83

So the issue seems to come from pkg-config? Anyone want to try building it and see if they spot what's happening?

`https://github.com/ar-jan/void-packages/tree/qgis-proj`

classabbyamp commented:
see #46285, i think i got everything working

ar-jan commented:
+1, works with QGIS, will update after grass is merged.

classabbyamp commented on srcpkgs/qgis/template:
```diff
@@ -0,0 +1,39 @@
+archs="x86_64* i686"
+build_style=cmake
+configure_args="-DENABLE_TESTS=OFF -DWITH_3D=True -DWITH_PDAL=False
+ -DWITH_GRASS7=False -DWITH_GRASS8=True
```
```suggestion
```
those aren't needed, they get set to those values automatically

ar-jan replied:
Removed

ar-jan commented:
One minor issue I encounter is that tooltips for toolbar icons are sometimes partially rendered, then usually rendering fully on the next hover. I have no clue why that is, I guess something Qt related. But basically I think it's ready.

nezos commented:
> One minor issue I encounter is that tooltips for toolbar icons are sometimes partially rendered, then usually rendering fully on the next hover. I have no clue why that is, I guess something Qt related. But basically I think it's ready.

I didn't compile the posted branch but toolbar icons in 3.32.2 render normaly here - running i3 on X (if this matters).

nezos commented:
One last comment, I think geotiff library is essential but we can look at it in a next update.

Good job!

ar-jan commented:
The icons do render ok for me, it's only the tooltips shown on hovering that that are sometimes incomplete for me on first hover (I'm running KDE Plasma).

Ah right geotiff would be good. I see you posted a template above, do you want to list yourself as maintainer?

nezos commented:
> The icons do render ok for me, it's only the tooltips shown on hovering that that are sometimes incomplete for me on first hover (I'm running KDE Plasma).
> 
> Ah right geotiff would be good. I see you posted a template above, do you want to list yourself as maintainer?

You can take ownership:) No problem! ... I mean if libgeotiff gets updated than QGIS will be updated too.

nezos commented:
Just wanted to add that as time passes, newer versions have more requirements. I am adding my latest build templates in case you find them useful. Another new library is being used by QGIS, draco, while 3 more are required for building. GRASS (latest version) is now been detected without the need for a patch and is shown properly in the tools.
Here are all my templates and a patch:

[draco.txt](https://github.com/void-linux/void-packages/files/13231442/draco.txt)
[freexl.txt](https://github.com/void-linux/void-packages/files/13231444/freexl.txt)
[qgis.txt](https://github.com/void-linux/void-packages/files/13231445/qgis.txt)
[grass.txt](https://github.com/void-linux/void-packages/files/13231449/grass.txt)
[libgeotiff.txt](https://github.com/void-linux/void-packages/files/13231454/libgeotiff.txt)
[libpdal.txt](https://github.com/void-linux/void-packages/files/13231457/libpdal.txt)
[libspatialite.txt](https://github.com/void-linux/void-packages/files/13231464/libspatialite.txt)
[libgdal.txt](https://github.com/void-linux/void-packages/files/13231469/libgdal.txt)
[libkml.txt](https://github.com/void-linux/void-packages/files/13231473/libkml.txt)
[libkml-NOUNCRYPT.patch.txt](https://github.com/void-linux/void-packages/files/13231475/libkml-NOUNCRYPT.patch.txt)
[python3-OWSLib.txt](https://github.com/void-linux/void-packages/files/13231478/python3-OWSLib.txt)
[python3-pyproj.txt](https://github.com/void-linux/void-packages/files/13231487/python3-pyproj.txt)
[python3-PyQt5-QScintilla.txt](https://github.com/void-linux/void-packages/files/13231490/python3-PyQt5-QScintilla.txt)

In case someone wants to build QGIS the sequence is:
1) libkml, 2) python3-OWSLib, 3) proj, python3-pyproj, 4) libgeotiff, 5) libgdal, 6) freexl, libspatialite, 7) libpdal, 8) grass, 9) python3-PyQt5-QScintilla, 10) qgis

ar-jan commented:
After the the move to Python 3.12 I had trouble with with PyQGIS, and I believe some core functionality broke (missing some menu items) just with rebuilding 3.32 for Python 3.12. Have you seen that? Haven't tried 3.34 yet, will take this into account, thanks!

nezos commented:
3.34 works fine and I also have the grass and gdal tools working. I don't see anything missing from the menus.

ar-jan commented:
Updated to 3.34. The apparent Python 3.12 issues are gone (grass had been working correctly for me since the 8.3 update). I've now also included PDAL.

classabbyamp commented on srcpkgs/draco/template:
```diff
@@ -0,0 +1,24 @@
+	pkg_install() {
+		vmove usr/include
+		vmove usr/lib/pkgconfig
+		vmove "usr/lib/*.a"
```
why only the static lib?

ar-jan replied:
The package has no `.so`s

classabbyamp replied:
on both of these packages with only .as, is there no configure option to enable shared libraries?

ar-jan replied:
Oh you're right, the generic cmake option works.

classabbyamp commented on common/shlibs:
generally new additions go at the bottom, I think

classabbyamp commented on srcpkgs/libgeotiff/template:
```diff
@@ -0,0 +1,29 @@
+	short_desc+=" - development files"
+	pkg_install() {
+		vmove usr/include
+		vmove "usr/lib/*.a"
```
why only static?

ar-jan replied:
Here too there are no `.so` libs.

ar-jan replied:
Changed to use `BUILD_SHARED_LIBS=ON` (excludes static archives)

classabbyamp commented on srcpkgs/libpdal/template:
```diff
@@ -0,0 +1,40 @@
+pkgname=libpdal
+version=2.5.6
+revision=1
+archs="x86_64* i686"
```
why only x86?

ar-jan replied:
I didn't get it to compile on arm because it uses some vendored tool `dimbuilder` on which it failed with `/bin/sh: line 1: dimbuilder: command not found`. It's also not entirely clear on which architectures it's [supposed to compile](https://github.com/PDAL/PDAL/issues/4239), though probably it should work on some arm arches. Any tips what to try there? Might `nocross` help here?

classabbyamp replied:
i think nocross is probably reasonable

classabbyamp commented on srcpkgs/qgis/template:
```diff
@@ -0,0 +1,38 @@
+pkgname=qgis
+version=3.34.0
+revision=1
+archs="x86_64* i686"
```
does this make more sense as nocross?

ar-jan replied:
I guess, I listed these because that's what other distros support for QGIS. Will give it a try.

classabbyamp commented on srcpkgs/libpdal/template:
```diff
@@ -0,0 +1,42 @@
+pkgname=libpdal
+version=2.5.6
+revision=1
+build_style=cmake
```
```suggestion
build_style=cmake
build_helper="qemu"
```

classabbyamp commented on srcpkgs/libpdal/template:
```diff
@@ -0,0 +1,42 @@
+homepage="https://pdal.io/"
+distfiles="https://github.com/PDAL/PDAL/releases/download/${version}/PDAL-${version}-src.tar.gz"
+checksum=c4f51e7bb9843ecc5d51c47aad87d8ddbd919111fb49222294c5835be7c1fa85
+nocross="dimbuilder: command not found"
```
```suggestion
```

classabbyamp commented:
QGIS will still be nocross due to grass, but getting grass to work on cross is much less trivial and i'm not going to bother with that for now
